File: pipeline/win_engine.py
#!/usr/bin/env python3
"""Win-total simulation engine (reference implementation).

MODEL (documented for owner evaluation)
=======================================
Ratings are neutral-field point margins vs an average FBS team (0). For a game between the
subject team S (rating mu_S) and opponent O (rating mu_O):

    expected_margin = mu_S - mu_O + HFA * site        site in {+1 home, -1 away, 0 neutral}
    P(S wins)       = Phi(expected_margin / sigma_eff)

This is a probit (normal-CDF) win model. Two variance sources, treated distinctly because
they correlate differently across a season:

1. GAME randomness (sigma_game): even with perfectly known ratings, a single game's margin
   scatters around its expectation. INDEPENDENT across games. Calibrated to CFB
   spread->win% conversion: sigma_game = 13.5 (a 7-pt favorite ~ Phi(7/13.5)=70%, a 3-pt
   favorite ~59%, a 14-pt favorite ~85% — matching how books price CFB moneylines).

2. OPPONENT rating uncertainty (opp band): we're unsure of each opponent's true rating.
   INDEPENDENT across the season's different opponents, so it adds to the per-game spread:
        sigma_eff_g = sqrt(sigma_game^2 + (band_O_g * BAND_TO_SD)^2)

3. OUR rating uncertainty (subject band): we're unsure of the SUBJECT team's true rating.
   This is a SHARED shock — if we're 2 pts too high, we're too high in EVERY game — so it
   is CORRELATED across all of the team's games. Modeled as a latent offset
        delta ~ Normal(0, tau_S^2),  tau_S = band_S * BAND_TO_SD
   applied to mu_S in every game, then integrated out. This shared shock is what fattens
   the win-total tails (extreme seasons become more likely than an independent-game model
   would say) — the single most important modeling choice here.

WIN DISTRIBUTION (exact, not Monte Carlo)
    Given delta, games are independent -> wins ~ Poisson-Binomial with probs p_g(delta),
    computed EXACTLY by DP (O(G^2)). Then mix over delta by Gauss-Hermite quadrature:
        P(wins=k) = sum_i w_i * PoissonBinomial(k | {p_g(delta_i)})
    with delta_i = sqrt(2)*tau_S*x_i and weights w_i/sqrt(pi) (x_i,w_i = Hermite nodes).
    21 nodes -> effectively exact for this smooth integrand; deterministic (no simulation
    noise). The same fixed nodes are shipped to the browser so the JS matches to 1e-9.

DEFAULTS (all adjustable, documented, exposed in the UI):
    HFA = 2.3, sigma_game = 13.5, BAND_TO_SD = 1.0 (band treated as ~1 SD of rating).
"""
import math

HFA = 2.3
SIGMA_GAME = 13.5
BAND_TO_SD = 1.0

# 21-point Gauss-Hermite nodes/weights (physicists', weight e^{-x^2}); hardcoded so the
# Python reference and the browser JS produce identical numbers. (numpy.hermgauss(21))
GH_NODES = [
 -5.550351873264678, -4.773992343411219, -4.121995547659471, -3.531972877137959,
 -2.979991207704598, -2.453552124512838, -1.944962949186254, -1.448934250650732,
 -0.961499634418369, -0.479450707079108, 0.0, 0.479450707079108, 0.961499634418369,
 1.448934250650732, 1.944962949186254, 2.453552124512838, 2.979991207704598,
 3.531972877137959, 4.121995547659471, 4.773992343411219, 5.550351873264678]
GH_WEIGHTS = [
 3.720365070136023e-14, 8.818611242049933e-11, 2.5712301800593154e-08, 2.17188489805667e-06,
 7.478398867310063e-05, 0.0012549820417264088, 0.011414065837434397, 0.0601796466589123,
 0.19212032406699775, 0.3816690736135022, 0.47902370312017756, 0.3816690736135022,
 0.19212032406699775, 0.0601796466589123, 0.011414065837434397, 0.0012549820417264088,
 7.478398867310063e-05, 2.17188489805667e-06, 2.5712301800593154e-08, 8.818611242049933e-11,
 3.720365070136023e-14]
_SQRT_PI = math.sqrt(math.pi)


def phi(z):
    """Standard normal CDF."""
    return 0.5 * (1.0 + math.erf(z / math.sqrt(2.0)))


def game_win_prob(mu_self, mu_opp, site, band_opp, sigma_game=SIGMA_GAME,
                  hfa=HFA, band_to_sd=BAND_TO_SD, delta=0.0):
    sigma_eff = math.sqrt(sigma_game * sigma_game + (band_opp * band_to_sd) ** 2)
    return phi((mu_self + delta - mu_opp + hfa * site) / sigma_eff)


def poisson_binomial(probs):
    """Exact distribution of the number of successes among independent Bernoulli(probs)."""
    dist = [1.0]
    for p in probs:
        p = min(max(p, 0.0), 1.0)
        nd = [0.0] * (len(dist) + 1)
        for k, dk in enumerate(dist):
            nd[k] += dk * (1.0 - p)
            nd[k + 1] += dk * p
        dist = nd
    return dist


def win_distribution(mu_self, band_self, games, sigma_game=SIGMA_GAME, hfa=HFA,
                     band_to_sd=BAND_TO_SD):
    """games: list of dicts {mu_opp, site, band_opp}. Returns dict with the mixed win
    distribution (list P(wins=k), k=0..G), expected wins, and the base p_g at delta=0."""
    tau = band_self * band_to_sd
    G = len(games)
    mixed = [0.0] * (G + 1)
    for x, w in zip(GH_NODES, GH_WEIGHTS):
        delta = math.sqrt(2.0) * tau * x
        probs = [game_win_prob(mu_self, g['mu_opp'], g['site'], g['band_opp'],
                               sigma_game, hfa, band_to_sd, delta) for g in games]
        pb = poisson_binomial(probs)
        wq = w / _SQRT_PI
        for k in range(G + 1):
            mixed[k] += wq * pb[k]
    s = sum(mixed)
    mixed = [m / s for m in mixed]                      # renormalize (guards tiny GH drift)
    exp_wins = sum(k * mixed[k] for k in range(G + 1))
    base_p = [game_win_prob(mu_self, g['mu_opp'], g['site'], g['band_opp'],
                            sigma_game, hfa, band_to_sd, 0.0) for g in games]
    return {'dist': mixed, 'expected_wins': exp_wins, 'base_p': base_p, 'G': G}


def ladder(dist):
    """P(wins >= k) for k=0..G+1 (survival ladder)."""
    G = len(dist) - 1
    out = {}
    for k in range(0, G + 2):
        out[k] = sum(dist[j] for j in range(k, G + 1))
    return out


# ---- American-odds utilities ----
def prob_to_american(p):
    p = min(max(p, 1e-12), 1 - 1e-12)
    return -100.0 * p / (1.0 - p) if p >= 0.5 else 100.0 * (1.0 - p) / p


def american_to_prob(a):
    return (-a) / ((-a) + 100.0) if a < 0 else 100.0 / (a + 100.0)


def american_to_decimal(a):
    return 1.0 + (a / 100.0 if a > 0 else 100.0 / (-a))


def under_from_over(over_american, cents=30):
    """Owner's convention: 30-cent line. over -175 <-> under +145."""
    return (-over_american - cents) if over_american < 0 else -(over_american + cents)


def market_edge(our_p_over, line, over_american, cents=30):
    """Returns dict of de-vigged market prob, our prob, edges, and EV per $1 each side."""
    under_american = under_from_over(over_american, cents)
    p_over_vig = american_to_prob(over_american)
    p_under_vig = american_to_prob(under_american)
    devig = p_over_vig + p_under_vig
    mkt_p_over = p_over_vig / devig
    our_p_under = 1.0 - our_p_over
    ev_over = our_p_over * (american_to_decimal(over_american) - 1) - (1 - our_p_over)
    ev_under = our_p_under * (american_to_decimal(under_american) - 1) - (1 - our_p_under)
    return {
        'line': line, 'over_odds': over_american, 'under_odds': under_american,
        'mkt_p_over': mkt_p_over, 'mkt_p_under': 1 - mkt_p_over, 'hold': devig - 1,
        'our_p_over': our_p_over, 'our_p_under': our_p_under,
        'edge_over': our_p_over - mkt_p_over, 'edge_under': our_p_under - (1 - mkt_p_over),
        'ev_over': ev_over, 'ev_under': ev_under,
    }

File: pipeline/win_engine.js
/* Win-total engine — JS port of pipeline/win_engine.py.
   Mirrors the Python reference exactly (same GH nodes/weights are injected from the payload).
   Normal CDF uses a Cephes ndtr port so phi() matches Python's math.erf to ~1e-14. */
(function (root) {
  'use strict';

  // ---- Cephes ndtr (normal CDF), double-precision ----
  function polevl(x, c) { var r = c[0]; for (var i = 1; i < c.length; i++) r = r * x + c[i]; return r; }
  function p1evl(x, c) { var r = x + c[0]; for (var i = 1; i < c.length; i++) r = r * x + c[i]; return r; }
  var T = [9.60497373987051638749e0, 9.00260197203842689217e1, 2.23200534594684319226e3,
           7.00332514112805075473e3, 5.55923013010394962768e4];
  var U = [3.35617141647503099647e1, 5.21357949780152679795e2, 4.59432382970980127987e3,
           2.26290000613890934246e4, 4.92673942608635921086e4];
  var P = [2.46196981473530512524e-10, 5.64189564831068821977e-1, 7.46321056442269912687e0,
           4.86371970985681366614e1, 1.96520832956077098242e2, 5.26445194995477358631e2,
           9.34528527171957607540e2, 1.02755188689515710272e3, 5.57535335369399327526e2];
  var Q = [1.32281951154744992508e1, 8.67072140885989742329e1, 3.54937778887819891062e2,
           9.75708501743205489753e2, 1.82390916687909736289e3, 2.24633760818710981792e3,
           1.65666309194161350182e3, 5.57535340817727675546e2];
  var R = [5.64189583547755073984e-1, 1.27536670759978104416e0, 5.01905042251180477414e0,
           6.16021097993053585195e0, 7.40974269950448939160e0, 2.97886665372100240670e0];
  var S = [2.26052863220117276590e0, 9.39603524938001434673e0, 1.20489539808096656605e1,
           1.70814450747565897222e1, 9.60896809063285878198e0, 3.36907645100081516050e0];
  var SQRTH = 7.07106781186547524401e-1, MAXLOG = 7.09782712893383996732e2;

  function erf(x) {
    if (Math.abs(x) > 1.0) return 1.0 - erfc(x);
    var z = x * x;
    return x * polevl(z, T) / p1evl(z, U);
  }
  function erfc(a) {
    var x = Math.abs(a);
    if (x < 1.0) return 1.0 - erf(a);
    var z = -a * a;
    if (z < -MAXLOG) return (a < 0) ? 2.0 : 0.0;
    z = Math.exp(z);
    var p, q;
    if (x < 8.0) { p = polevl(x, P); q = p1evl(x, Q); }
    else { p = polevl(x, R); q = p1evl(x, S); }
    var y = (z * p) / q;
    if (a < 0) y = 2.0 - y;
    if (y === 0.0) return (a < 0) ? 2.0 : 0.0;
    return y;
  }
  function ndtr(a) {  // normal CDF
    var x = a * SQRTH, z = Math.abs(x);
    if (z < SQRTH) return 0.5 + 0.5 * erf(x);
    var y = 0.5 * erfc(z);
    return (x > 0) ? (1.0 - y) : y;
  }

  function phi(z) { return ndtr(z); }

  // ---- engine (mirrors win_engine.py) ----
  function makeEngine(meta) {
    var HFA = meta.hfa, SIGMA_GAME = meta.sigma_game, BAND_TO_SD = meta.band_to_sd;
    var NODES = meta.gh_nodes, WEIGHTS = meta.gh_weights;
    var SQRT_PI = Math.sqrt(Math.PI), SQRT2 = Math.sqrt(2.0);

    function gameWinProb(muSelf, muOpp, site, bandOpp, opts) {
      opts = opts || {};
      var sg = opts.sigma_game != null ? opts.sigma_game : SIGMA_GAME;
      var hfa = opts.hfa != null ? opts.hfa : HFA;
      var bts = opts.band_to_sd != null ? opts.band_to_sd : BAND_TO_SD;
      var delta = opts.delta != null ? opts.delta : 0.0;
      var sigmaEff = Math.sqrt(sg * sg + Math.pow(bandOpp * bts, 2));
      return phi((muSelf + delta - muOpp + hfa * site) / sigmaEff);
    }

    function poissonBinomial(probs) {
      var dist = [1.0];
      for (var i = 0; i < probs.length; i++) {
        var p = Math.min(Math.max(probs[i], 0.0), 1.0);
        var nd = new Array(dist.length + 1).fill(0.0);
        for (var k = 0; k < dist.length; k++) {
          nd[k] += dist[k] * (1.0 - p);
          nd[k + 1] += dist[k] * p;
        }
        dist = nd;
      }
      return dist;
    }

    // games: [{mu_opp, site, band_opp}]
    function winDistribution(muSelf, bandSelf, games, opts) {
      opts = opts || {};
      var sg = opts.sigma_game != null ? opts.sigma_game : SIGMA_GAME;
      var hfa = opts.hfa != null ? opts.hfa : HFA;
      var bts = opts.band_to_sd != null ? opts.band_to_sd : BAND_TO_SD;
      var tau = bandSelf * bts, G = games.length;
      var mixed = new Array(G + 1).fill(0.0);
      for (var n = 0; n < NODES.length; n++) {
        var delta = SQRT2 * tau * NODES[n];
        var probs = games.map(function (g) {
          return gameWinProb(muSelf, g.mu_opp, g.site, g.band_opp,
            { sigma_game: sg, hfa: hfa, band_to_sd: bts, delta: delta });
        });
        var pb = poissonBinomial(probs);
        var wq = WEIGHTS[n] / SQRT_PI;
        for (var k = 0; k <= G; k++) mixed[k] += wq * pb[k];
      }
      var s = mixed.reduce(function (a, b) { return a + b; }, 0);
      mixed = mixed.map(function (m) { return m / s; });
      var ew = 0; for (var k2 = 0; k2 <= G; k2++) ew += k2 * mixed[k2];
      var baseP = games.map(function (g) {
        return gameWinProb(muSelf, g.mu_opp, g.site, g.band_opp,
          { sigma_game: sg, hfa: hfa, band_to_sd: bts, delta: 0.0 });
      });
      return { dist: mixed, expected_wins: ew, base_p: baseP, G: G };
    }

    return { gameWinProb: gameWinProb, poissonBinomial: poissonBinomial,
             winDistribution: winDistribution, phi: phi,
             HFA: HFA, SIGMA_GAME: SIGMA_GAME, BAND_TO_SD: BAND_TO_SD };
  }

  // ---- odds utilities (mirror win_engine.py) ----
  function probToAmerican(p) {
    p = Math.min(Math.max(p, 1e-12), 1 - 1e-12);
    return p >= 0.5 ? (-100.0 * p / (1.0 - p)) : (100.0 * (1.0 - p) / p);
  }
  function americanToProb(a) { return a < 0 ? (-a) / ((-a) + 100.0) : 100.0 / (a + 100.0); }
  function americanToDecimal(a) { return 1.0 + (a > 0 ? a / 100.0 : 100.0 / (-a)); }
  function underFromOver(over, cents) {
    cents = cents == null ? 30 : cents;
    return over < 0 ? (-over - cents) : -(over + cents);
  }

  var api = { makeEngine: makeEngine, phi: phi, erf: erf, erfc: erfc, ndtr: ndtr,
              probToAmerican: probToAmerican, americanToProb: americanToProb,
              americanToDecimal: americanToDecimal, underFromOver: underFromOver };
  if (typeof module !== 'undefined' && module.exports) module.exports = api;
  root.WinEngine = api;
})(typeof self !== 'undefined' ? self : this);
